lib/container.ts

```diff
@@ -0,0 +1,251 @@
+import * as Bluebird from 'bluebird';
+import * as Dockerode from 'dockerode';
+import * as _ from 'lodash';
+import { fs } from 'mz';
+import * as path from 'path';
+import * as Stream from 'stream';
+import * as tar from 'tar-stream';
+
+import Dockerfile, { DockerfileActionGroup } from './dockerfile';
+import { ContainerNotRunningError, InternalInconsistencyError } from './errors';
+import { streamToBuffer } from './util';
+
+/**
+ * This structure represents files which have changed that are
+ * part of the build context. They should be provided relative to
+ * the build context itself, for example:
+ * 	build context: /usr/src/app
+ * 	changed file: /usr/src/app/src/main.ts
+ * should be provided as:
+ * 	src/main.ts
+ */
+export interface FileUpdates {
+	updated: string[];
+	added: string[];
+	deleted: string[];
+}
+
+interface AddOperation {
+	fromPath: string;
+	toPath: string;
+}
+type DeleteOperation = string;
+
+export interface ExecResult {
+	returnCode: number;
+	stdoutOutput: Buffer;
+	stderrOutput: Buffer;
+}
+
+export class Container {
+	private dockerfile: Dockerfile;
+
+	public constructor(
+		public dockerfileContent: string,
```

Comment:
We can't `readonly` the `containerId` as this can (and will in future PRs) be changed. This is why we repeatedly make a call to `docker.getContainer(this.containerId)` (although this could be encapsulated inside a getter/setter, we'll see). Agreed for the rest, and pretty sure `docker` can be, but it can be changed if need be.